### PR TITLE
CAMBI heatmap generation

### PR DESCRIPTION
**Pull request by nilfm**
* Enables string options for VMAF
* Factors out buffers used by CAMBI from `CambiState`  into `CambiBuffers` to avoid clutter
* Adds optional parameter to CAMBI which provides a path to save one 16-bit YUV400 (grayscale) video for each scale, representing the `c_values` at that scale. This allows the user to visualize the bands detected by CAMBI and their temporal movement.

Comments:
* If we agree that the YUV400 option is better than the multiple PFM option, we can squash the last 4 commits into one. I'm keeping it this way for now in case we want to leave the last commit out (in which case, we can squash the last 3).
* It'd be good to test this on Windows, I don't have a Windows system but I'll try to find one to try it on. Mostly worried about the path separator.
* I've tested this on Mac OS by opening the generated .yuv files with `vooya`.
* (minor) I'm not sure if `.yuv` is the right extension for single-channel videos, I haven't found a better one.
* It would be good to check if the provided path is not empty and exit to avoid possibly corrupting user data.

### CODE REVIEW

**Comment by kylophone** [comment]
This will probably have merge conflicts with #941.

File: libvmaf/src/feature/cambi.c [hunk 1]
```diff
@@ -719,9 +734,33 @@ static FORCE_INLINE inline double weight_scores_per_scale(double *scores_per_sca
     return score / normalization;
 }
 
-static int cambi_score(VmafPicture *pics, uint32_t *mask_dp, uint16_t window_size, double topk,
-                       const uint16_t *tvi_for_diff, float *c_values, uint16_t *c_values_histograms, 
-                       uint8_t *filter_mode_histogram, uint16_t *filter_mode_buffer, double *score) {
+void write_uint16(FILE *file, uint16_t v) {
```
Comment (ltrudeau-twoorioles) [inline libvmaf/src/feature/cambi.c:737]: This function could probably be `static`.

File: libvmaf/src/feature/cambi.c [hunk 2]
```diff
@@ -719,9 +734,33 @@ static FORCE_INLINE inline double weight_scores_per_scale(double *scores_per_sca
...
+        }
+    }
+    char path[1024];
+    snprintf(path, sizeof(path), "%s/cambi_heatmap_scale_%d.yuv", dir_path, scale);
+    int max_c_value = max_diff_weight * window_size * window_size / 4;
+    int scaling_value = (1 << 16) - 1;
+    FILE *file = fopen(path, "a");
+    for (int i = 0; i < height; i++) {
+        for (int j = 0; j < width; j++) {
+            uint16_t val = scaling_value * c_values[i * width + j] / max_c_value;
```
Comment (ltrudeau-twoorioles) [inline libvmaf/src/feature/cambi.c:755]: Might be faster to compute `max_c_value_inv = 1/max_c_value` outside the nested loops and use `* max_c_value_inv` inside the nested loops

File: libvmaf/src/feature/cambi.c [hunk 3]
```diff
@@ -719,9 +734,33 @@ static FORCE_INLINE inline double weight_scores_per_scale(double *scores_per_sca
...
 }
 
-static int cambi_score(VmafPicture *pics, uint32_t *mask_dp, uint16_t window_size, double topk,
-                       const uint16_t *tvi_for_diff, float *c_values, uint16_t *c_values_histograms, 
-                       uint8_t *filter_mode_histogram, uint16_t *filter_mode_buffer, double *score) {
+void write_uint16(FILE *file, uint16_t v) {
+    fwrite((void*)(&v), sizeof(v), 1, file);
+}
+
+static void dump_c_values(float *c_values, int width, int height, int scale, int window_size, char *dir_path) {
```
Comment (ltrudeau-twoorioles) [inline libvmaf/src/feature/cambi.c:741]: How about `const float *c_values` and `const char *dir_path`

**Comment by ltrudeau-twoorioles** [comment]
> * (minor) I'm not sure if `.yuv` is the right extension for single-channel videos, I haven't found a better one.

How about `.y4m`?

File: libvmaf/src/feature/cambi.c (hunk 2 above)
Comment (nilfm) [reply libvmaf/src/feature/cambi.c:755]: True, we can even do `scaling_value = ((1 << 16) - 1)/max_c_value` since we're multiplying by that anyway.

**Comment by nilfm** [comment]
@ltrudeau-twoorioles, I have discussed the appropriate file extension with @krasuluk and we found that `.gray` is used elsewhere for this purpose, so we'll be using that. Thanks for the comments!

File: libvmaf/src/feature/cambi.c [hunk 4]
```diff
@@ -719,9 +734,34 @@ static FORCE_INLINE inline double weight_scores_per_scale(double *scores_per_sca
...
+        if (g_diffs_weights[i] > max_diff_weight) {
+            max_diff_weight = g_diffs_weights[i];
+        }
+    }
+    char path[1024];
+    snprintf(path, sizeof(path), "%s/cambi_heatmap_scale_%d_%dx%d_16b.gray", dir_path, scale, width, height);
+    int max_c_value = max_diff_weight * window_size * window_size / 4;
+    int max_16bit_value = (1 << 16) - 1;
+    double scaling_value = (double)max_16bit_value / max_c_value;
+    FILE *file = fopen(path, "a");
```
Comment (kylophone) [inline libvmaf/src/feature/cambi.c:753]: Your call to `fopen()` can fail, add a check here and bail if it's no good.

File: libvmaf/src/feature/cambi.c [hunk 5]
```diff
@@ -719,9 +734,34 @@ static FORCE_INLINE inline double weight_scores_per_scale(double *scores_per_sca
...
+
+static void dump_c_values(const float *c_values, int width, int height, int scale, int window_size, const char *dir_path) {
+    int max_diff_weight = g_diffs_weights[0];
+    for (int i = 0; i < NUM_DIFFS; i++) {
+        if (g_diffs_weights[i] > max_diff_weight) {
+            max_diff_weight = g_diffs_weights[i];
+        }
+    }
+    char path[1024];
+    snprintf(path, sizeof(path), "%s/cambi_heatmap_scale_%d_%dx%d_16b.gray", dir_path, scale, width, height);
```
Comment (kylophone) [inline libvmaf/src/feature/cambi.c:749]: "/" is probably not a cross-platform.

File: libvmaf/src/feature/cambi.c (hunk 4 above)
Comment (nilfm) [reply libvmaf/src/feature/cambi.c:753]: By "bail" you mean abort the whole program, right? I would say that continuing execution without writing to disk is not desirable.

Comment (kylophone) [reply libvmaf/src/feature/cambi.c:753]: No, just return an error and handle it somehow. If fopen fails and your file pointer is NULL you will get a SIGSEGV.

File: libvmaf/src/feature/cambi.c (hunk 5 above)
Comment (nilfm) [reply libvmaf/src/feature/cambi.c:749]: I have seen some [posts](https://stackoverflow.com/questions/12971499/how-to-get-the-file-separator-symbol-in-standard-c-c-or/30261289) suggesting something like this to set up the separator:
```
const char separator = 
#ifdef _WIN32
    '\\'
#else
    '/'
#endif
;
```

There are a couple references to `_WIN32` in this repo, but none of them is used to determine the separator in this way. Do you think this is fine?

**Comment by nilfm** [comment]
I'll wait for #941 to be pushed and then rebase these changes on top of that.

**Comment by kylophone** [comment]
#941 is merged. This one can be rebased now.

**Comment by nilfm** [comment]
Rebased on top of #941 changes

**Review by joelsole — APPROVED** [review]

**Comment by XuShiSheng** [comment]
你好！邮件已收到，我会尽快回复你--许思焱

File: libvmaf/src/feature/cambi.c [hunk 6]
```diff
@@ -911,6 +920,11 @@ static int close_cambi(VmafFeatureExtractor *fex) {
     aligned_free(s->buffers.mask_dp);
     aligned_free(s->buffers.filter_mode_histogram);
     aligned_free(s->buffers.filter_mode_buffer);
+
+    for (int scale = 0; scale < NUM_SCALES; scale++) {
```
Comment (kylophone) [inline libvmaf/src/feature/cambi.c:924]: maybe add a guard for s->heatmaps_path here.

**Pull request merged**